### PR TITLE
Fix silent error when reporting builds to CDash

### LINKED ISSUE #47723: CDash Reporting not working as expected with long build names
CDash reports that contain build errors do not display on CDash when a long build name is passed to spack for the report.  This can be replicated by building a package that has build errors with the following command:

```command
spack install --cdash-track=track --cdash-upload-url=a-very-long...url <package>
```
where "a-very-long...url" is a string containing at least 256 characters.  In this case Spack creates a Build.xml file containing the build name and then uploads that XML to the provided CDash URL by first encoding the CDash track and CDash build name into the upload URL.  It appears that the CDash server truncates the build name as encoded in the URL to  255 characters, but the build name in the XML is still full-length.  This creates a build name miss-match between the build name as reported in the URL and the build name as reported in the XML.  When this happens the report looks like it was submitted correctly and the build line shows up on CDash with the correct number of errors.  However, when attempting to view the specifics of the build errors by clicking on them, a blank page containing only the following text is displayed:
```text
Site: test
Build Name: a-very-long...url
Build Time: 2024-11-21 14:37:44
Found 0 Errors
Warnings are here. 
```
It is relatively easy to generate long build names when using spack.  In my case this occurred as a result of using the definition of the spec (with all variants specified) for each package being reported.

I'm wondering what the best approach to fixing this is.  Should Spack do something to assert a max length for build names or perhaps this is an issue for Kitware to look into CDash side?  @zackgalbreath can you provide any insight into what the right approach might be here?

### PR DESCRIPTION
**Pull request by jnhealy2**
CDash has a 191 char maximum for build names.  When this is exceeded, CDash silently fails to correctly process the reported XML. This truncates CDash build names to 190 chars and emits a warning indicating it is doing so to prevent such errors from occuring.

closes #47723

### CODE REVIEW

**Comment by jnhealy2** [comment]
@psakievich can you please add the snl-core-team label?

**Review by psakievich — DISMISSED** [review]

**Review by tldahlgren — CHANGES_REQUESTED** [review]
LGTM except it needs a unit test.  

Not sure how much help you want. I recommend adding a test to `test/reporters.py`.  You can instantiate `CDashConfiguration` with fake data where `build` exceeds the length; instantiate `CDash` with the configuration; and confirm that the results of calling the modified method yields a name that doesn't match that of the configuration and or is the truncated length.

**Review by psakievich — DISMISSED** [review]
LGTM. Deferring merging to give @tldahlgren a chance to review the added test.

File: lib/spack/spack/ci/common.py [hunk 1]
```diff
@@ -351,7 +351,7 @@ def report_skipped(self, spec: spack.spec.Spec, report_dir: str, reason: Optiona
         configuration = CDashConfiguration(
             upload_url=self.upload_url,
             packages=[spec.name],
-            build=self.build_name,
+            build=self.build_name(),
```
Comment (tldahlgren) [inline lib/spack/spack/ci/common.py:354]: Line 220's use of the method also needs `()`.

File: lib/spack/spack/reporters/cdash.py [hunk 2]
```diff
@@ -123,11 +124,15 @@ def __init__(self, configuration: CDashConfiguration):
...
 
     def report_build_name(self, pkg_name):
-        return (
+        buildname = (
             "{0} - {1}".format(self.base_buildname, pkg_name)
             if self.multiple_packages
             else self.base_buildname
         )
+        if len(buildname) > 190:
+            warnings.warn("Build name exceeds CDash 191 character maximum and will be truncated.")
```
Comment (tldahlgren) [inline lib/spack/spack/reporters/cdash.py]: The limit is technically 190.

```suggestion
            warnings.warn("Build name exceeds CDash 190 character maximum and will be truncated.")
```

**Review by tldahlgren — CHANGES_REQUESTED** [review]
LGTM. A couple of minor fixes in the comments.

File: lib/spack/spack/ci/common.py (hunk 1 above)
Comment (jnhealy2) [reply lib/spack/spack/ci/common.py:354]: should be taken care of now, thanks for pointing this out @tldahlgren

**Review by tldahlgren — APPROVED** [review]
Thanks.

**Review by psakievich — APPROVED** [review]

**Pull request merged**